src/api/providers/fireworks.ts

```diff
@@ -0,0 +1,222 @@
+			}
+			return
+		}
+
+		// Format the messages according to OpenAI format, with special handling for DeepSeek models
+		let openAiMessages: OpenAI.Chat.ChatCompletionMessageParam[] = []
+		const isDeepSeek = model.id.includes("deepseek")
+
+		if (isDeepSeek) {
+			// DeepSeek models handle system prompts differently
+			if (systemPrompt.trim() !== "") {
+				openAiMessages.push({ role: "system", content: systemPrompt })
+			}
+			openAiMessages = [...openAiMessages, ...convertToOpenAiMessages(messages)]
+		} else {
+			// Standard OpenAI format
+			openAiMessages = [{ role: "system", content: systemPrompt }, ...convertToOpenAiMessages(messages)]
+		}
+
+		// Calculate token usage for prompt size limits
+		let estimatedTokens = 0
+		const contextWindow = model.info.contextWindow || 8191
+
+		try {
+			const contentBlocks = openAiMessages.map((msg) => ({
+				text: typeof msg.content === "string" ? msg.content : JSON.stringify(msg.content),
+			}))
+
+			estimatedTokens = await this.countTokens(contentBlocks as any)
+		} catch (error) {
+			// Silently handle token counting errors
+		}
+
+		// Calculate max output tokens based on context window and input size
+		const maxOutputTokens = Math.min(model.info.maxTokens || 4096, Math.max(100, contextWindow - estimatedTokens))
+
+		try {
+			// Prepare request parameters
+			const requestOptions: OpenAI.Chat.Completions.ChatCompletionCreateParamsStreaming = {
+				model: model.id,
+				messages: openAiMessages,
+				max_tokens: maxOutputTokens,
+				temperature: isDeepSeek ? 0.7 : 0.6, // DeepSeek works better with slightly higher temperature
+				top_p: 1,
+				presence_penalty: 0,
+				frequency_penalty: 0,
+				stream: true,
+				stream_options: { include_usage: true },
+			}
+
+			// Send the request
+			const stream = await this.client.chat.completions.create(requestOptions)
+
+			// Process the streaming response
+			let contentStarted = false
+			let accumulatedText = ""
+
+			for await (const chunk of stream) {
+				const delta = chunk.choices[0]?.delta
+				if (delta?.content) {
+					contentStarted = true
+					accumulatedText += delta.content
+					yield {
+						type: "text",
+						text: delta.content,
+					}
+				}
+
+				// Process usage information if available
+				if (chunk.usage) {
+					yield {
+						type: "usage",
+						inputTokens: chunk.usage.prompt_tokens || 0,
+						outputTokens: chunk.usage.completion_tokens || 0,
+						totalCost: calculateApiCostOpenAI(
+							model.info,
+							chunk.usage.prompt_tokens || 0,
+							chunk.usage.completion_tokens || 0,
+						),
+					}
+				}
+			}
+
+			// Handle the case where no content was received
+			if (!contentStarted) {
+				yield {
+					type: "text",
+					text: "No content was received from the model. This could be an issue with the API or the model.",
+				}
+			}
+		} catch (error: any) {
+			// Handle specific error cases with user-friendly messages
+			if (error.status === 401) {
+				yield {
```

Comment:
Similar to my comment above about the API key, I don't see any of the other providers do this. Is this AI slop? What happens with those other providers if they hit e.g. a rate limit?

Reply:
Actually this is a great point. I would be worth quickly checking if the SDK does this better. That would save us a bunch of custom code. You can just load the SDK in a nodejs REPL and see what errors it gives.